File: C/heapsort.c
#include "<stdio.h>"

int create_heap(int[16] array){
	int heap_pointer = 0;
	int ptr;
	int new_ptr;
	bool stop;
	int temp;
	for (heap_pointer = 0;heap_pointer<16;heap_pointer++){
		ptr = heap_pointer;
		stop = false;
		while ((ptr <> 0)&&(stop = false)){
			new_ptr = (ptr-1)/2
			if (array[new_ptr]<array[ptr]){
				temp = array[new_ptr];
				array[new_ptr] = array[ptr];
				array[ptr] = temp;
			}
			ptr = new_ptr;
		}
	}
	return array;
}

int sort_heap(int[16] array){
	int heap_pointer;
	int temp;
	int ptr;
	for (heap_pointer = 15;heap_pointer>=0;heap_pointer--){
		temp = array[0];
		ptr = 0;
		while (ptr <= heap_pointer){
			if(array[(2*ptr)+1]>array[(2*ptr)+2]){
				if((2*ptr)+1)<=heap_pointer){
					array[ptr] = array[(2*ptr)+1];
				}
				ptr = (2*ptr)+1;
			}
			else:
				if((2*ptr)+2<=heap_pointer){
					array[ptr] = array[(2*ptr)+2];
				}
				ptr = (2*ptr)+1;

		}
		array[heap_pointer] = temp;
	}
	return array;
}

void print_array(int[16] array){
	for(int i = 0;i<16;i++){
		printf("%i",array[i]);
	}
	printf("\n");
	return void;
}



int main(){

	int[16] heap;

	heap = [0,1,2,3,4,5,6,7,8,9,20,11,12,13,14,15];
	heap = create_heap(heap);
	print_array(heap);
	heap = sort_heap(heap);
	print_array(heap);
	return 0;
}

int create_heap(int[16] array){
	int heap_pointer = 0;
	int ptr;
	int new_ptr;
	bool stop;
	int temp;
	for (heap_pointer = 0;heap_pointer<16;heap_pointer++){
		ptr = heap_pointer;
		stop = false;
		while ((ptr <> 0)&&(stop = false)){
			new_ptr = (ptr-1)/2
			if (array[new_ptr]<array[ptr]){
				temp = array[new_ptr];
				array[new_ptr] = array[ptr];
				array[ptr] = temp;
			}
			ptr = new_ptr;
		}
	}
	return array;
}

int sort_heap(int[16] array){
	int heap_pointer;
	int temp;
	int ptr;
	for (heap_pointer = 15;heap_pointer>=0;heap_pointer--){
		temp = array[0];
		ptr = 0;
		while (ptr <= heap_pointer){
			if(array[(2*ptr)+1]>array[(2*ptr)+2]){
				if((2*ptr)+1)<=heap_pointer){
					array[ptr] = array[(2*ptr)+1];
				}
				ptr = (2*ptr)+1;
			}
			else:
				if((2*ptr)+2<=heap_pointer){
					array[ptr] = array[(2*ptr)+2];
				}
				ptr = (2*ptr)+1;

		}
		array[heap_pointer] = temp;
	}
	return array;
}

void print_array(int[16] array){
	for(int i = 0;i<16;i++){
		printf("%i",array[i]);
	}
	printf("\n");
	return void;
}

File: Python/Misc/heapsort.py
class Heap(list):
	def __init__(self,array):
		for heap_pointer in xrange(len(array)):
			print array
			ptr = heap_pointer
			stop = False
			while (not stop) and (ptr != 0):
				new_ptr = (ptr-1)/2
				if array[new_ptr]<array[ptr]: #if child is greater than parent
					array[ptr],array[new_ptr] = array[new_ptr],array[ptr] #swap
				ptr = new_ptr
		self.heap = array

	def sort(self):
		array =self.heap
		value_range = list(xrange(len(array)))
		value_range.reverse()
		for heap_pointer in value_range:
			print array
			temp = array[0]
			ptr = 0
			while ptr <= heap_pointer:
				try:
					if array[(2*ptr)+1]>array[(2*ptr)+2]:
						if (2*ptr)+1<=heap_pointer:
							array[ptr] = array[(2*ptr)+1]
						ptr = (2*ptr)+1
					else:
						if (2*ptr)+2<=heap_pointer:
							array[ptr] = array[(2*ptr)+2]
						ptr = (2*ptr)+2
				except IndexError:
					try:
						if (2*ptr)+1<=heap_pointer:
							array[ptr] = array[(2*ptr)+1]
						break
					except:
						break
			array[heap_pointer] = temp
		return array

print "creating heap"
Heap = Heap([0,4,2,4,6,8,7,2,9,1])
print "\n\n\nSorting heap"
print Heap.sort()
